nat-lab/tests/test_events_link_state.py

```diff
@@ -0,0 +1,412 @@
+    "alpha_setup_params",
+    [
+        pytest.param(
+            SetupParameters(
+                connection_tag=ConnectionTag.DOCKER_CONE_CLIENT_1,
+                adapter_type=AdapterType.LinuxNativeWg,
+                connection_tracker_config=generate_connection_tracker_config(
+                    ConnectionTag.DOCKER_CONE_CLIENT_1,
+                    derp_1_limits=ConnectionLimits(1, 1),
+                ),
+                features=TelioFeatures(
+                    link_detection=LinkDetection(rtt_seconds=1),
+                    wireguard=long_persistent_keepalive_periods(),
+                ),
+            )
+        ),
+    ],
+)
+@pytest.mark.parametrize(
+    "beta_setup_params",
+    [
+        pytest.param(
+            SetupParameters(
+                connection_tag=ConnectionTag.DOCKER_CONE_CLIENT_2,
+                adapter_type=AdapterType.LinuxNativeWg,
+                connection_tracker_config=generate_connection_tracker_config(
+                    ConnectionTag.DOCKER_CONE_CLIENT_2,
+                    derp_1_limits=ConnectionLimits(1, 1),
+                ),
+                features=TelioFeatures(
+                    link_detection=LinkDetection(rtt_seconds=1),
+                    wireguard=long_persistent_keepalive_periods(),
+                ),
+            )
+        ),
+    ],
+)
+async def test_event_link_state_peer_goes_offline(
+    alpha_setup_params: SetupParameters, beta_setup_params: SetupParameters
+) -> None:
+    async with AsyncExitStack() as exit_stack:
+        env = await setup_mesh_nodes(
+            exit_stack, [alpha_setup_params, beta_setup_params]
+        )
+        alpha, beta = env.nodes
+        client_alpha, client_beta = env.clients
+        connection_alpha, connection_beta = [
+            conn.connection for conn in env.connections
+        ]
+
+        async with Ping(connection_alpha, beta.ip_addresses[0]).run() as ping:
+            await testing.wait_long(ping.wait_for_next_ping())
+        async with Ping(connection_beta, alpha.ip_addresses[0]).run() as ping:
+            await testing.wait_long(ping.wait_for_next_ping())
+
+        await client_beta.stop_device()
+
+        await asyncio.sleep(1)
+
+        with pytest.raises(asyncio.TimeoutError):
+            async with Ping(connection_alpha, beta.ip_addresses[0]).run() as ping:
+                await testing.wait_normal(ping.wait_for_next_ping())
+
+        await asyncio.sleep(15)
+        alpha_events = client_beta.get_link_state_events(alpha.public_key)
+        beta_events = client_alpha.get_link_state_events(beta.public_key)
+
+        # 1 down when node is Connecting and 1 up when node is Connected
+        assert alpha_events == [LinkState.Down, LinkState.Up]
+        # beta will have 2 down events: 1 when is Connecting and 1 detected
+        assert beta_events == [LinkState.Down, LinkState.Up, LinkState.Down]
+
+
+# TODO: Add the scenario where peers exchange data then idle then resume
```

Comment:
Will this be addressed in a follow-up PR?

Reply:
Yes, for sure.
Also the comments left unresolved from this PR.